server/src/main/java/org/elasticsearch/index/mapper/MapperService.java

```diff
@@ -119,12 +121,20 @@ public MapperService(IndexSettings indexSettings, IndexAnalyzers indexAnalyzers,
         super(indexSettings);
         this.indexVersionCreated = indexSettings.getIndexVersionCreated();
         this.indexAnalyzers = indexAnalyzers;
-        this.documentParser = new DocumentParser(xContentRegistry);
-        this.documentMapperParser = new DocumentMapperParser(indexSettings, this, similarityService, mapperRegistry,
-                queryShardContextSupplier, scriptService);
         this.indexAnalyzer = new MapperAnalyzerWrapper(indexAnalyzers.getDefaultIndexAnalyzer(), MappedFieldType::indexAnalyzer);
         this.mapperRegistry = mapperRegistry;
-        this.idFieldDataEnabled = idFieldDataEnabled;
+        Function<DateFormatter, Mapper.TypeParser.ParserContext> parserContextFunction =
+            dateFormatter -> new Mapper.TypeParser.ParserContext(similarityService::getSimilarity, mapperRegistry.getMapperParsers()::get,
+                indexVersionCreated, queryShardContextSupplier, dateFormatter, scriptService, indexAnalyzers, indexSettings,
+                idFieldDataEnabled);
+        this.documentParser = new DocumentParser(xContentRegistry, parserContextFunction);
+        this.parserContextSupplier = () -> new Mapper.TypeParser.ParserContext(similarityService::getSimilarity,
+            mapperRegistry.getMapperParsers()::get, indexVersionCreated,
+            queryShardContextSupplier, null, scriptService, indexAnalyzers, indexSettings, idFieldDataEnabled);
```

Comment:
Can we merge these two by just calling the function with `null` if there's no date formatter?